### DISCUSSION TITLE
How to use javet in multi-threaded environment with heavy load?

### DISCUSSION THREAD

**Discussion by Lookout-ent** [caoccao/Javet#234] [Q&A]
I have a script, in which i need to invoke a function with different input to function each time. i want to use javet to create v8runtime environment in java to invoke the javascript function and i need to do this in multi-threaded environment and for a longer period of time , basically load is very high.
what is the best way to achieve this ?

Note: 1) I tried creating a new runtime environment for every thread, which results in memory spike and basically application is crashing
2) I created a single runtime environment and used it for every threads. In my junit for 10-20 threads, i didn't see it crashing. But in my application, it's crashing. Which could be because of memory spike.

I'm very new to using javet.Please guide me through this.

**Comment by caoccao** [comment]
1. The crashing usually is caused by the memory leak. I guess there are flaws in your implementation. Please read the doc (best practices) and review the Javet unit test for some ideas on how to achieve zero memory leak.
2. I recommend the Javet engine pool for your use case. It works like a DB connection pool. You just let Javet engine pool manage the pooling algo and focus on your own stuff.

**Reply by Lookout-ent** [reply]
Yes it seems there are some memory leak 
javet engine pool is a great thing
i was wondering if javet provides any way to pre-compile the script into a binary format and loading it into all engines in the pool.
i want to avoid the overhead of parsing and compiling the script each time an engine is acquired from the pool.

**Reply by caoccao** [reply]
Yes, Javet supports loading pre-compiled functions or modules from a byte array.

But, if you use the pool, I don't think that's necessary. I recommend the following design.

1. Wrap all the shared code in modules.
2. Compile the modules while the engine is being created by sub-classing the built-in Javet engine.
3. Wrap the actual business logic in a module inside which use dynamic import to import those shared modules.
4. Execute that module with the actual business logic.

The shared module compilation only takes place per engine during the initialization, so that there's no need to save to or load from a byte array.
